insights: added documentation for the historical backfiller (#22332)
sourcegraph/sourcegraph: doc/dev/background-information/insights/backend.md

It is a moderately complex algorithm, to get an understanding for how it operates see these two explanations:

1. [what it does and general implementation thoughts](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@049b99763b10ddc7ef6f72c22b18f1fa5f4f7259/-/blob/enterprise/internal/insights/background/historical_enqueuer.go#L31-53)
2. [overview of the algorithm](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/insights/background/historical_enqueuer.go?L185#L116-147)

> Naively the historical backfiller will likely take a long time on any reasonably sized Sourcegraph installation. As an optimization,

Reviewer: ```suggestion
Naively implemented, the historical backfiller would take a long time on any reasonably sized Sourcegraph installation. As an optimization,
```

Once the web client gets data points back, it renders them! Contact @felixfbecker for details on where/how that happens.

These queries can be executed concurrently by using the site setting `insights.query.worker.concurrency` and providing
the desired concurrency factor. With `insights.query.worker.concurrency=1` queries will be executed in serial.

There is a rate limit associated with the query worker. This limit is shared across all concurrent handlers and can be configured

> using the site setting `insights.query.worker.rateLimit`. This value to set will depend on the size and scale of your Sourcegraph

Reviewer: ```suggestion
using the site setting `insights.query.worker.rateLimit`. This value to set will depend on the size and scale of the Sourcegraph
```

Reviewer: These docs are just for devs at Sourcegraph, doubt any customer will read them / stumble across them.
